src/Sarif/Visitors/InsertOptionalDataVisitor.cs

```diff
@@ -0,0 +1,148 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.Sarif.Visitors
+{
+    public class InsertOptionalDataVisitor : SarifRewritingVisitor
+    {
+        internal IFileSystem s_fileSystem = new FileSystem();
+
+        private Run _run;
+        private bool _overwriteExistingData;
+        private FileRegionsCache _fileRegionsCache;
+        private OptionallyEmittedData _dataToInsert;
+        
+        public InsertOptionalDataVisitor(OptionallyEmittedData dataToInsert, bool overwriteExistingData = true)
+        {
+            _dataToInsert = dataToInsert;
+            _overwriteExistingData = dataToInsert.Includes(OptionallyEmittedData.OverwriteExistingData);
+        }
+
+        public override Run VisitRun(Run node)
+        {
+            _run = node;
+
+            if (node != null)
+            {
+                bool scrapeFileReferences = _dataToInsert.Includes(OptionallyEmittedData.Hashes)      ||
+                                            _dataToInsert.Includes(OptionallyEmittedData.TextFiles)   ||
+                                            _dataToInsert.Includes(OptionallyEmittedData.BinaryFiles);
+
+                if (scrapeFileReferences)
+                {
+                    var visitor = new AddFileReferencesVisitor();
+                    visitor.VisitRun(node);
```

Comment:
>visitor.VisitRun(node) [](start = 20, length = 22)

Oh my! That's very clever (and at the same time quite clear, unlike most "clever" code).